### PR TITLE
Auth to companies controller 436

### PR DESCRIPTION
**Pull request by vijayakarumudi**
Closes AgileVentures/MetPlus_tracker#436

### CODE REVIEW

**Comment by vijayakarumudi** [comment]
PR for review

File: app/policies/company_policy.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  def destroy?
+    agency_admin?(user, record)
+  end
+
+  def show?
+    edit?
+  end
+
+  def list_people?
+    company_admin?(user, record) || agency_admin?(user, record) || user.is_company_person?(record)
```
Comment (joaopapereira) [inline app/policies/company_policy.rb]: Maybe the check for company admin is redundant if we will allow all the company people to see this

File: spec/controllers/branches_controller_spec.rb [hunk 2]
```diff
@@ -306,3 +307,6 @@
     end
   end
 end
+
```
Comment (joaopapereira) [inline spec/controllers/branches_controller_spec.rb]: Do we need to do changes in this file? If not can you please revert it?

File: spec/controllers/companies_controller_spec.rb [hunk 3]
```diff
@@ -105,4 +168,136 @@
...
+        allow(controller).to receive(:current_user).and_return(company_admin)
+        get :edit, id: company
+        expect(subject).to_not receive(:user_not_authorized)
+      end
+      it 'authorizes agency admin' do
+        allow(controller).to receive(:current_user).and_return(admin)
+        get :edit, id: company
+        expect(subject).to_not receive(:user_not_authorized)
+      end
+      it_behaves_like 'unauthorized agency people and jobseeker' do
```
Comment (joaopapereira) [inline spec/controllers/companies_controller_spec.rb]: Missing unauthorized company people

File: spec/controllers/companies_controller_spec.rb [hunk 4]
```diff
@@ -105,4 +168,136 @@
...
+        let(:my_request) { get :edit, id: company }
+      end
+    end
+    context '#update' do
+      it 'authorizes company admin' do
+        allow(controller).to receive(:current_user).and_return(company_admin)
+        params_hash = attributes_for(:company,
```
Comment (joaopapereira) [inline spec/controllers/companies_controller_spec.rb]: We can dry out the parameters to a before each block

File: spec/controllers/companies_controller_spec.rb [hunk 5]
```diff
@@ -105,4 +168,136 @@
...
+                                     email: 'info@widgets.com', fax: '510 555-1212',
+                                     job_email: 'humanresources@widgets.com')
+                      .merge(addresses_attributes:
+                         { '0' => attributes_for(:address),
+                           '1' => attributes_for(:address) })
+
+        patch :update, id: company, company: params_hash
+        expect(subject).to_not receive(:user_not_authorized)
+      end
+      it_behaves_like 'unauthorized agency people and jobseeker' do
```
Comment (joaopapereira) [inline spec/controllers/companies_controller_spec.rb]: We are missing unauthorized company people

File: spec/controllers/companies_controller_spec.rb [hunk 6]
```diff
@@ -105,4 +168,136 @@
...
+        get :show, id: company
+        expect(subject).to_not receive(:user_not_authorized)
+      end
+
+      it 'authorizes agency admin' do
+        allow(controller).to receive(:current_user).and_return(admin)
+        get :show, id: company
+        expect(subject).to_not receive(:user_not_authorized)
+      end
+      it_behaves_like 'unauthorized agency people and jobseeker' do
```
Comment (joaopapereira) [inline spec/controllers/companies_controller_spec.rb]: Missing unauthorized company people

File: spec/controllers/branches_controller_spec.rb (hunk 2 above)
Comment (vijayakarumudi) [inline spec/controllers/branches_controller_spec.rb]: yes. I have merge conflicts that why I made changes.

File: spec/controllers/companies_controller_spec.rb (hunk 6 above)
Comment (vijayakarumudi) [inline spec/controllers/companies_controller_spec.rb]: I fixed this one.

File: spec/controllers/companies_controller_spec.rb (hunk 4 above)
Comment (vijayakarumudi) [inline spec/controllers/companies_controller_spec.rb]: I dry out the parameter in this method

File: spec/controllers/companies_controller_spec.rb [hunk 7]
```diff
@@ -105,4 +168,136 @@
...
+        expect(subject).to_not receive(:user_not_authorized)
+      end
+
+      it_behaves_like 'unauthorized agency people and jobseeker' do
+        let(:my_request) { delete :destroy, id: company }
+      end
+
+      it_behaves_like 'unauthorized all' do
+        let(:my_request) { delete :destroy, id: company }
+      end
```
Comment (patmbolger) [inline spec/controllers/companies_controller_spec.rb]: Also missing unauthorized company contact.

File: app/controllers/companies_controller.rb [hunk 8]
```diff
@@ -35,17 +46,19 @@ def update
   end
 
   def list_people
-    raise 'Unsupported request' if not request.xhr?
```
Comment (vijayakarumudi) [inline app/controllers/companies_controller.rb]: I did not drop this line  actually.  I have to check  my code. by accident it happned

File: spec/controllers/companies_controller_spec.rb (hunk 7 above)
Comment (vijayakarumudi) [inline spec/controllers/companies_controller_spec.rb]: I already updated all tests for company contact.

**Comment by vijayakarumudi** [comment]
close PR

**Comment by patmbolger** [comment]
Code changes look good.  I think this is ready to merge.

**Comment by vijayakarumudi** [comment]
Updated PR

File: spec/controllers/branches_controller_spec.rb [hunk 9]
```diff
@@ -252,10 +253,16 @@
         expect(subject).to_not receive(:user_not_authorized)
       end
       it_behaves_like 'unauthorized all' do
-        let(:my_request) { post :create, agency_id: agency, branch: FactoryGirl.attributes_for(:branch) }
+        let(:my_request) do
```
Comment (joaopapereira) [inline spec/controllers/branches_controller_spec.rb]: do we need to change this file?

Comment (vijayakarumudi) [inline spec/controllers/branches_controller_spec.rb]: yes!. other wise rubocop found the offences detected.

**Comment by joaopapereira** [comment]
Please rebase and we can merge

**Comment by vijayakarumudi** [comment]
I updated PR.

**Pull request merged**